spec/functional/run_lock_spec.rb

```diff
@@ -267,20 +311,120 @@ def results
       Process.waitpid2(p1)
     end
 
-    it "doesn't truncate the lock file so that contents can be read" do
-      p1 = fork do
-        run_lock.acquire
-        run_lock.save_pid
-        sleep 2
-        exit! 1
+  end
+
+  class ClientProcess
+    def initialize(example, name)
+      @example = example
+      @name = name
+      @read_from_process, @write_to_tests = IO.pipe
+      @read_from_tests, @write_to_process = IO.pipe
+    end
+
+    attr_reader :example
+    attr_reader :name
+    attr_reader :pid
+
+    def event(event)
+      # Let the caller know what event we've reached
+      write_to_tests.puts(event)
+
+      # Block until the client tells us where to stop
+      if !@run_to_event || event == @run_to_event
+        @run_to_event = read_from_tests.gets.strip
       end
+    end
 
-      wait_on_lock
-      sleep 0.5 # Possible race condition on Solaris which pid is observed as 0
-      expect(File.read(lockfile)).to eq(p1.to_s)
+    def last_event
+      while true
+        event = readline_nonblock(read_from_process)
+        break if event.nil?
+        @last_event = event.strip
+      end
+      @last_event
+    end
 
-      Process.waitpid2(p1)
+    def run_to(event, &background_block)
+      # Start the process if it's not started
+      start if !pid
+
+      # Tell the process what to stop at (also means it can go)
+      write_to_process.puts event
+
+      # Run the background block
+      background_block.call if background_block
+
+      # Wait until it gets there
+      Timeout::timeout(CLIENT_PROCESS_TIMEOUT) do
+        until last_event == event
+          @last_event = read_from_process.gets.strip
```

Comment:
I'm confused about needing to do both blocking and non-blocking calls here.  If you aren't guaranteed to get the event, then this doesn't protect you that much.  You still need to worry about calling read_from_process.gets after last_event has returned because the pipe was closed.  If you are guaranteed to get the event, the last_event is unnecessary.  Just make this last_event a local variable initialized to nil and do blocking calls.